### DISCUSSION TITLE
Example of using v2 Parcel API (rebundling) to run dev-server?

### DISCUSSION THREAD

**Discussion by tajo** [parcel-bundler/parcel#4889] [General] [answered]
I want to you use [Parcel v2 API](https://v2.parceljs.org/features/parcel-api/) as a part of my bundle tool. It's pretty straightforward for the prod build (just running `bundler.run()`) but not sure how to set up a dev server.

Should I try to replicate this [function](https://github.com/parcel-bundler/parcel/blob/v2/packages/core/parcel/src/cli.js#L166)?

**Comment by DeMoorJasper** [comment, marked as answer]
If you just want to add some kind of middleware to the devserver you can utilise the proxy config.

If you want to write a completely custom server however, you'd probably want to make a custom reporter for that. This is the default dev-server reporter: https://github.com/parcel-bundler/parcel/tree/v2/packages/reporters/dev-server

**Reply by egriff38** [reply]
So is there no way to have a simple bundler middleware [as in version 1](https://parceljs.org/api.html#middleware)?

**Reply by DeMoorJasper** [reply]
I can't really answer that without knowing the use-case, as far as I know proxies will cover most use-cases and probably be easier than in Parcel 1

**Reply by egriff38** [reply]
I was just looking for a solution that wouldn't involve starting 2 top-level node server processes. Plus I found using a monorepo structure for the server and client and using the client as a direct dependency to be pretty slick ([example repo](https://github.com/egriff38/parcel-1-monorepo)).

**Reply by DeMoorJasper** [reply]
Still not sure what the benefit would be over just using the proxy in your example. You start express with only the parcel instance in the example, you could just start parcel and it'll do all that for you by default...

If you need to mount a backend server or something you can use the proxy feature and mount `/api` to the backend server or code. I really don't see any difference over using proxies for your use-case, although I might completely misunderstand

**Reply by egriff38** [reply]
The thing I've omitted here is authentication. I need a session token from the OAuth service I use to access the web pages. Routing on the page takes advantage to already being authenticated. In the past, I did a 2-way proxy:
- An express server served on `443` does authentication and has `/api` routes, along with a proxy to
- A parcel (or webpack) dev server hosting static assets on `8080` with an `/api` proxy back to `443`

Now for this, several things were privy to bugs:
- CORS issues between the 2 hosts
- Hot reloading broken
- Proxy path matching (I work in an environment with subpaths necessary to emulate production environment)
- My team had trouble understanding they needed to connect to the 443 server instead of the 8080 to see auth work correctly. 

These are the practical issues I've seen repeatedly with the Vue CLI specifically, and there's too much overhead in the webpack configs for me to eject the dev server there as a middleware. Parcel 1 offered a simple, intuitive node API for hooking into an existing web server.

I've only come here to inquire as I've seen some issues with sourcemaps not being able to compile are being addressed in Parcel 2, and I'm trying to understand whether there's a direct parallel between my current setup as in my example repo and the way the Parcel 2 API works.

As a side note, thank you so much for talking with me about this! I love the parcel ecosystem! 🥇

**Reply by DeMoorJasper** [reply]
@egriff38 this is totally possible with proxies, made an example here: https://github.com/DeMoorJasper/parcel-2-middleware-example

The `.proxyrc.js` file gets an express instance passed in so you can attach as many middlewares and logic as you want, it's actually the opposite of Parcel 1, now you create the middlewares instead of the server... Which seems easier

**Reply by egriff38** [reply]
Okay, so I didn't know the proxy was capable of same-process router proxying, thats super useful. This could work potentially if you already have the server logic available to you, by importing it into the `.proxyrc.js` file. I understand what you mean when you say it's the opposite. Technically, this could still work with the structure I shared, I would just need to import the backend routing into the frontend proxy. This stands in contrast to the way the project would be structured at build time, where the express server hosts the static assets produced by `parcel build` as opposed to dev runtime where parcel is itself holding pieces of the express routing. I understand the transition is under most cases transparent, however, it still offers some challenges over the old middleware model
- debugging the express server with the `--inspect` flag
- ~injecting `ts-node` or `dotenv` via the `-r` flag.~ Could be required at the top of the `.proxyrc.js` file.
- What is parcel's restart policy if a server file is modified? Does it restart the server or are dependencies of `.proxyrc.js` actually monitored? Should something like nodemon be used with parcel for just the server files?

**Reply by DeMoorJasper** [reply]
Dependencies of proxyrc are not watched, you probably don't even need it, you can load the server code on each request which removes the need for anything like nodemon as well. It will result in slightly longer request times...

`--inspect` is not really possible... as that will include all parcel code in this case there's no way to isolate this as far as I know. That's still the same as Parcel 1 I assume?

If you truly want all these things, you probably don't want to run parcel as any kind of middleware or your api as middleware, you probably want to run parcel in watch mode and just have your server/backend serve these files statically. https://v2.parceljs.org/features/cli/#parcel-watch-%3Centries%3E this way Parcel won't conflict with your backend in any way, you have full freedom and experience the same performance and experience as you would in production during development

**Reply by egriff38** [reply]
I was able to use the inspect flag if I imported the parcel 1 middleware, because the top level process was still started with `node` directly. I have used parcel in watch mode and honestly I agree with you that it's an easy while effective solution. The major benefit of transitioning from watch mode to the parcel 1 API import was that the parcel process and the express process shared `process.env`, which meant if I imported or set an environment variable on the server, the same changes would go through in the parcel process that inlines those variables (super helpful in debugging). 

I might reinvestigate the API strategy once the docs are more developed.

**Reply by DeMoorJasper** [reply]
Got ya we're also moving away from inlining environment variables without explicitly defining them as it's a pretty large security risk, so that ease of use might actually change in the future as well for security reasons... we're probably add some flag to enable it again though, a lot of frameworks and tools already do this as it becomes a pretty big issue once you have strict security requirements

About using `--inspect` you can still do that using `node node_modules/bin/parcel --inspect ...` if I'm not mistaken but it will be bloated with all kinds of Parcel code.

**Reply by mischnic** [reply]
I usually do `PARCEL_WORKERS=0 NODE_OPTIONS=--inspect yarn parcel .... `, though I'm not sure if that's actually helpful in this case
